core/src/main/scala/org/apache/spark/util/collection/ExternalAppendOnlyMap.scala

```diff
@@ -192,12 +193,16 @@ class ExternalAppendOnlyMap[K, V, C](
    * It will be called by TaskMemoryManager when there is not enough memory for the task.
    */
   override protected[this] def forceSpill(): Boolean = {
-    assert(readingIterator != null)
-    val isSpilled = readingIterator.spill()
-    if (isSpilled) {
-      currentMap = null
+    if (isReadingIterator) {
+      assert(readingIterator != null)
+      val isSpilled = readingIterator.spill()
+      if (isSpilled) {
+        currentMap = null
+      }
+      isSpilled
+    } else {
+      false
```

Comment:
As I commented before:

> I choose to simply return false now. Another option is to actually spill the in-memory map.

If `readingIterator` is not set, once `forceSpill` is called we can spill the content of `currentMap` into disk like `spill` does now.

We can't just instantiate `readingIterator` and spill it. Because `readingIterator` is only created for iterating over the entries of the in-memory map, when we want to read the content of this `ExternalAppendOnlyMap`. If `forceSpill` is called before we want to read the content, and we instantiate `readingIterator` and spill it, the in-memory map will be released and we can't insert data further.